src/rules/noUnnecessaryCallbackWrapperRule.ts

```diff
@@ -44,40 +44,42 @@ export class Rule extends Lint.Rules.AbstractRule {
     }
 }
 
-function walk(ctx: Lint.WalkContext<void>) {
-    return ts.forEachChild(ctx.sourceFile, cb);
-    function cb(node: ts.Node): void {
-        if (isArrowFunction(node) && !hasModifier(node.modifiers, ts.SyntaxKind.AsyncKeyword) &&
-            isCallExpression(node.body) && isIdentifier(node.body.expression) &&
-            isRedundantCallback(node.parameters, node.body.arguments)) {
-            const start = node.getStart(ctx.sourceFile);
-            ctx.addFailure(start, node.end, Rule.FAILURE_STRING(node.body.expression.text), [
-                Lint.Replacement.deleteFromTo(start, node.body.getStart(ctx.sourceFile)),
-                Lint.Replacement.deleteFromTo(node.body.expression.end, node.end),
+function walk(ctx: Lint.WalkContext<void>): void {
+    const { sourceFile } = ctx;
+    return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        const call = detectUnnecessaryCallback(node);
+        if (call !== undefined) {
+            const start = node.getStart(sourceFile);
+            ctx.addFailure(start, node.end, Rule.FAILURE_STRING(call.expression.getText()), [
+                Lint.Replacement.deleteFromTo(start, call.getStart(sourceFile)),
+                Lint.Replacement.deleteFromTo(call.expression.end, node.end),
             ]);
         } else {
             return ts.forEachChild(node, cb);
         }
-    }
-
+    });
 }
 
-function isRedundantCallback(parameters: ts.NodeArray<ts.ParameterDeclaration>, args: ts.NodeArray<ts.Node>): boolean {
-    if (parameters.length !== args.length) {
-        return false;
+function detectUnnecessaryCallback(node: ts.Node): ts.CallExpression | undefined {
+    if (!isArrowFunction(node) || hasModifier(node.modifiers, ts.SyntaxKind.AsyncKeyword)) {
+        return undefined;
+    }
+    const { parameters, body } = node;
+    if (!isCallExpression(body)) {
+        return undefined;
     }
-    for (let i = 0; i < parameters.length; ++i) {
-        const {dotDotDotToken, name} = parameters[i];
+    const { expression, arguments: args } = body;
+    const isRedundant = isIdentifier(expression) && parameters.length === args.length && parameters.every(({ dotDotDotToken, name }, i) => {
         let arg = args[i];
         if (dotDotDotToken !== undefined) {
             if (!isSpreadElement(arg)) {
                 return false;
             }
             arg = arg.expression;
         }
-        if (!isIdentifier(name) || !isIdentifier(arg) || name.text !== arg.text) {
-            return false;
-        }
-    }
-    return true;
+        return isIdentifier(name) && isIdentifier(arg) && name.text === arg.text
+            // If the invoked expression is one of the parameters, bail.
+            && expression.text !== name.text;
```

Comment:
this is the only line that really changed
please consider keeping the diff small and just add this line to the source as it is on master.
Btw. sorry for the rather big refactor and therefore unnecessary big diff in #2510 :confused: